### PR TITLE
Add new node services watch function

### PR DESCRIPTION
**Pull request by jorgemarey**
This PR adds a new function to the watch package.
The functions allows to watch new for new services on a node.

### CODE REVIEW

**Comment by i0rek** [comment]
Thank you for your work @jorgemarey! Are you still interested in adding this?

**Comment by jorgemarey** [comment]
Hi @i0rek. Yes, I'm still interested. I see that the files had chaged location to api/watch. I should made the changes there, right?

**Comment by i0rek** [comment]
@jorgemarey your feature seems reasonable to me, but could you help me understand your use case better just to make sure we are on the same page? When we agree on the adding this feature I will review your PR.

**Comment by jorgemarey** [comment]
I developed a simple tcp connect proxy but that runs at a node level. It watches all node_services and for each one of them that needs a sidecar, stars a new port and listen there. Pretty simple.

The code is something like this:

```go
// ConfigWatcher is a simple interface to allow dynamic configurations from
// pluggable sources.
type ConfigWatcher interface {
	// Watch returns a channel that will deliver new Configs if something external
	// provokes it.
	Watch() <-chan *Config
}

// NodeServicesConfigWatcher watches the local Consul node for proxy service changes.
type NodeServicesConfigWatcher struct {
	....
}

// NewNodeServicesConfigWatcher creates an AgentConfigWatcher.
func NewNodeServicesConfigWatcher(client *api.Client,
	logger *log.Logger) (*NodeServicesConfigWatcher, error) {
	...
	// Setup watch plan for config
	plan, err := watch.Parse(map[string]interface{}{
		"type": "node_services",
		"node": nodeName,
	})
	if err != nil {
		return nil, err
	}
	w.plan = plan
	w.plan.HybridHandler = w.handler
	go w.plan.RunWithClientAndLogger(w.client, w.logger)
	return w, nil
}

func (w *NodeServicesConfigWatcher) handler(blockVal watch.BlockingParamVal, val interface{}) {
	....
        configs := make(map[string]*proxy.Config)
	for _, svc := range resp.Services {
		if svc.Kind != api.ServiceKindConnectProxy {
			continue
		}
                ...
		cfg := &proxy.Config{
			ProxiedServiceName:      svc.Proxy.DestinationServiceName,
			ProxiedServiceNamespace: "default",
                        ...
		}
		configs[svc.ID] = cfg
	}
	w.ch <- &Config{
		proxyConfigs: configs,
	}
}

...

// Serve the proxy instance until a fatal error occurs or proxy is closed.
func (p *Proxy) Serve() error {
	for {
		select {
		case cfg := <-p.cfgWatcher.Watch():
			p.logger.Printf("[DEBUG] got new config")
			for id, config := range cfg.proxyConfigs {
				p.runServiceProxy(id, config) // this starts a new proxy if it wasn't configured previously
			}
			p.removeNotPresent(cfg.proxyConfigs) // stops removed proxies
		case <-p.stopChan:
			return nil
		}
	}
}
```

I guess that it would make more sense to use an agent/service endpoint, but it doesn't allow blocking queries I developed this one.

Let me know if this makes sense to you. Thanks!

**Comment by i0rek** [comment]
Thank you for your explanation! `agentServiceWatch` is not what you need since it is for a single service. But what do you think about adding `agentServicesWatch` which would support blocking? Your use case is to query for services for the local agent and no catalog request has to be made for this.

(Explainer: I am still trying to figure out what would be the best solution for your problem)

**Comment by jorgemarey** [comment]
According to the documentation the endpoint agent/services do not support blocking queries: https://www.consul.io/api/agent/service.html#list-services

**Comment by i0rek** [comment]
Right! How could I miss that endpoint!? Well, I propose that we add the ability to block on that query then with `LocalBlockingQuery`. Thoughts?

**Comment by jorgemarey** [comment]
Yep. That makes sense. I'll need to change the agent/services endpoint [here](https://github.com/hashicorp/consul/blob/401221de5852432d742f2bdb1c4f17fb0d0e7362/agent/agent_endpoint.go#L212) to use `LocalBlockingQuery` as in [here](https://github.com/hashicorp/consul/blob/401221de5852432d742f2bdb1c4f17fb0d0e7362/agent/agent_endpoint.go#L291). Right? I'll start with the changes as soon as I can. Should I made another different PR for this?

**Comment by i0rek** [comment]
Yes sounds good! You can either use this PR or create a new one. Both things are fine.

File: api/agent.go [hunk 1]
```diff
@@ -415,6 +415,28 @@ func (a *Agent) ServicesWithFilter(filter string) (map[string]*AgentService, err
 	return out, nil
 }
 
+func (a *Agent) ServicesWithQueryOptions(filter string, q *QueryOptions) (map[string]*AgentService, *QueryMeta, error) {
```
Comment (jorgemarey) [inline api/agent.go]: I created a new function to avoid a breaking change in the API package.

File: agent/agent_endpoint.go [hunk 2]
```diff
@@ -227,29 +234,68 @@ func (s *HTTPServer) AgentServices(resp http.ResponseWriter, req *http.Request)
 		return nil, err
 	}
 
-	services := s.agent.State.Services(&entMeta)
-	if err := s.agent.filterServicesWithAuthorizer(authz, &services); err != nil {
-		return nil, err
-	}
+	wch := make(chan struct{})
```
Comment (jorgemarey) [inline agent/agent_endpoint.go]: The documentation states that state.Notify sends a message on the channel whenever a state change occur. It says that for now only service registration and deregistration, but that could change. I don't know if using this here is the best option.

**Comment by jorgemarey** [comment]
Hi @i0rek, some tests are failiing, but I'm not sure if they are related to this changes.

**Comment by jsosulska** [comment]
Hello @jorgemarey ,

Thank you for your patience with our responses on this PR. We've gone ahead and rebased the issue, and are waiting on CI. Can you please just reply here once you've reviewed the changes, or if there are additions you would like to make.

Thank you again for your contribution to this feature!
Best,
Jono

File: agent/agent_endpoint_test.go [hunk 3]
```diff
@@ -6215,3 +6216,172 @@ func TestAgent_Services_ExposeConfig(t *testing.T) {
...
+		},
+		{
+			name:     "blocking fetch timeout, no change",
+			url:      "/v1/agent/services?hash=2593138afc0803c4&wait=100ms",
+			wantWait: 100 * time.Millisecond,
+			wantCode: 200,
+			wantResp: expectedResponse1,
+		},
+		{
+			name:     "blocking fetch old hash should return immediately",
```
Comment (dnephin) [inline agent/agent_endpoint_test.go:6319]:
```suggestion
			name:     "blocking fetch with an expired hash should return immediately",
```

File: agent/agent_endpoint_test.go [hunk 4]
```diff
@@ -6215,3 +6216,172 @@ func TestAgent_Services_ExposeConfig(t *testing.T) {
...
+			resp := httptest.NewRecorder()
+			if tt.updateFunc != nil {
+				go tt.updateFunc()
+			}
+			start := time.Now()
+			obj, err := a.srv.AgentServices(resp, req)
+			elapsed := time.Since(start)
+
+			fmt.Println(resp.HeaderMap)
+
```
Comment (dnephin) [inline agent/agent_endpoint_test.go:6358]: Looks like some left over debugging. This is causing the linter to complain

```suggestion
```

**Review by dnephin — COMMENTED** [review]
Thank you for this PR!

It looks like the test for the new `api` method is failing. We had a quick look, but couldn't locate the problem. It looks like for some reason the  ContentHash header is not being set.  I think once that test is passing this change should be ready to merge.

**Comment by dnephin** [comment]
It looks like this PR has not received any updates in over a year, so I'm going to close it.

If you are still interested in updating this PR, please comment and I can re-open it, or opening a new PR would also work.

Thank you again for your interest in contributing this enhancement to Consul!

**Pull request closed without merging**